workspace/update-injected-packages/src/index.ts

```diff
@@ -0,0 +1,59 @@
+import path from 'path'
+import { fetchFromDir } from '@pnpm/directory-fetcher'
+import { PnpmError } from '@pnpm/error'
+import { type ImportOptions, createIndexedPkgImporter } from '@pnpm/fs.indexed-pkg-importer'
+import { readCurrentLockfile } from '@pnpm/lockfile.fs'
+import { globalInfo } from '@pnpm/logger'
+import { findInjectedPackages } from './findInjectedPackages'
+
+export interface UpdateInjectedPackagesOptions {
+  pkgName: string | undefined
+  pkgRootDir: string
+  virtualStoreDir: string | undefined
+  virtualStoreDirMaxLength: number
+  workspaceDir: string | undefined
+}
+
+export async function updateInjectedPackages (opts: UpdateInjectedPackagesOptions): Promise<void> {
+  if (!opts.pkgName) {
+    globalInfo(`Skip updating ${opts.pkgRootDir} as an injected package because without name, it cannot be a dependency`)
+    return
+  }
+  if (!opts.virtualStoreDir) {
+    throw new PnpmError('NO_VIRTUAL_STORE_DIR', 'Cannot update injected packages without virtual store dir')
+  }
+  if (!opts.workspaceDir) {
+    throw new PnpmError('NO_WORKSPACE_DIR', 'Cannot update injected packages without workspace dir')
+  }
+  const lockfile = await readCurrentLockfile(opts.virtualStoreDir, {
+    ignoreIncompatible: false,
+  })
+  if (!lockfile) {
+    globalInfo('Stop updating injected packages because no current lockfile means no dependency')
+    return
+  }
+  const pkgRootDir = path.resolve(opts.workspaceDir, opts.pkgRootDir)
+  const items = [...findInjectedPackages({
+    lockfile,
+    pkgName: opts.pkgName,
+    pkgRootDir,
+    workspaceDir: opts.workspaceDir,
+    virtualStoreDir: opts.virtualStoreDir,
+    virtualStoreDirMaxLength: opts.virtualStoreDirMaxLength,
+  })]
+  if (items.length === 0) {
+    globalInfo(`Found no injected packages matching ${opts.pkgRootDir}`)
+    return
+  }
+  const { filesIndex } = await fetchFromDir(pkgRootDir, {})
+  const importOptions: ImportOptions = {
+    filesMap: filesIndex,
+    force: true,
+    resolvedFrom: 'local-dir',
+  }
+  const importPackage = createIndexedPkgImporter('clone-or-copy')
+  for (const { targetDir } of items) {
+    globalInfo(`Importing ${targetDir} from ${pkgRootDir}`)
+    importPackage(targetDir, importOptions)
```

Comment:
I am not sure if this is the best way to do it. This will basically override the target directory instead of adding the missing files. It would be better just to extend the existing directory. Although that wouldn't be an atomic operation, which might be OK

Reply:
I think it should have the same effect as a clean reinstall (with lockfile).

Furthermore, injected packages, when deduped will become symbolic links. If symlinked packages and deduped injected workspace packages don't retain files that have been deleted at the source, why behave differently when it's non-deduped injected?

Reply:
I wasn't saying that deleted files should be retained. Deleted files should be removed.

Reply:
Then I guess your concern was performance?

If we can read all the directories and compare it with the source to know which files to deleted, retained, and added. This would minimize the number of syscalls at the cost of complexity.

If we just overwrite, the number of syscalls would increase but the code would be very simple. The syscalls are all `mkdir` and `link` which should be relatively cheap (I don't know whether `createIndexedPkgImporter('hardlink')` would fallback to copying files, but even if it does, there are few filesystems that people use that don't support hardlinks).

The time complexity of both approaches should be O(n) but with different values of n.

Reply:
It can be done in a separate PR but in my opinion it makes sense to do. And I don't think that it is hard to implement. Yes, you'll have to make a lot of read operations but aren't read operations less stressful for the filesystem than a lot of unneeded write operations? We could even compare the contents of files and only replace them if they differ. In this case, I think only a few write operations will be needed after a compilation.

Reply:
> We could even compare the contents of files and only replace them if they differ.

They are hardlinks. Comparing won't be necessary.

Reply:
For hard links yes. But what if the dist directory is recreated from scratch? In that case those are new files, some of which might be the same as the old ones.

Reply:
OK. We can compare inodes to check if they are links. If they aren't, just copy overwrite because I believe 1 call to `fs.copyFile` isn't more expensive than reading 2 a whole files into memory (in JS) then compare them (in JS).

Reply:
OK, this should be fine as we will basically cleanup. Even if the files are equal, they use additional disk space, so replacing them with a hard link sounds justified. But we need to use hard links not copyFile.

Reply:
> But we need to use hard links not copyFile.

I was typing it half-sleep.

Reply:
I don't think indexed-pkg-importer needs diffing as the packages imported to node_modules are always added atomically to empty directories.